php-lover/assignment11/app/api/calculate.py

```diff
@@ -0,0 +1,32 @@
+import functools
+from app.helper.calc import Calculator
+
+from flask import request, make_response
+from app.helper.db import get_db
+from app.api import api_bp
+
+
+@api_bp.route('/calculate', methods=['POST'])
+def calculate():
+    calc = Calculator()
+    data = request.json
+
+    try:
+        equation = data['to_calculate']
+        result = calc.calculate(equation)
+    except KeyError:
+        return make_response({"error": "Missing required parameter 'to_caclulate'"}, 400)
+    except Exception as e:
+        return make_response({"error": "Could not process equation", "reason": str(e)}, 400)
+
+    try:
+        db = get_db()
+        db.execute(
+            "INSERT INTO calculate_history (input, result) VALUES (?, ?)",
+            (equation, result)
+        )
+        db.commit()
```

Comment:
I would recommend using the Flask ORM here, it's called SQLAlchemy. Pretty sure this logic would be vulnerable to an SQL Injection which is the kind of thing an ORM protects you against.

Reply:
SQLAlchemy would also encourage you to use a models.py file as well as the makemigrations -> migrate flow which is best practice.

Reply:
This is a parameterized statement, so there is no injection vulnerability here. Not how the SQL statement has `?` placeholders and the actual variables are given as an additional parameter to the function call.

As for SQLAlchemy, I deliberately chose not to include it due to the size and scope of the project. Part of using best practices is understanding when they're not actually the best for your use case, and in this instance including an ORM for a single table with three fields would add tons of unnecessary additional overhead.